[Workers] (WIP) Assets (#16610)
cloudflare/cloudflare-docs: src/content/docs/workers/static-assets/compat-matrix.mdx

---
pcx_content_type: concept

> title: Compatibility Matrix

Reviewer: Especially if this is going to be under Static Assets, seems more specific to say:

```suggestion
title: Workers vs Pages (compatibility matrix)
```

or something that names workers and pages specifically in title, in nav.

otherwise I read this as "this is a compatibility matrix of static assets" which I don't follow

sidebar:
  order: 10
head: []
description: Compatibility matrix for asset hosting on Cloudflare Workers and Pages.
---

import {
	Badge,
	Description,
	FileTree,
	InlineBadge,
	Render,
	TabItem,
	Tabs,
} from "~/components";

> Cloudflare [Workers with static assets](/workers/static-assets/) and [Pages](/pages/) both have the ability to host frontend applications. The following compatibility matrix shows the features are available for each product to help you when choosing which product to start with.

Reviewer: ```suggestion
You can deploy frontend applications to both Cloudflare [Workers with static assets](/workers/static-assets/) and [Pages](/pages/), from static websites to full-stack apps. The following compatibility matrix shows which are currently available for each product to help you when choosing which product to start with.
```

> We plan to bridge the gaps between Workers and Pages and provide ways to migrate your Pages projects to Workers.

Reviewer: ```suggestion
We are working to bridge gaps between Workers and Pages by bringing Pages-only features to Workers, and Workers-only features to Pages. And we will provide ways to easily migrate Pages projects to Workers.
```

| --------------------------------- | ------------------------ | ----- |
| **Features**                      |                          |       |
| Middleware                        | ❌ <sup>1</sup>          | ✅    |
| Redirects                         | ❌ <sup>2</sup>          | ✅    |
| Smart Placement                   | ❌ <sup>3</sup>          | ✅    |
| **Observability**                 |                          |       |
| Persistent Logs                   | ✅                       | ❌    |
| Logpush                           | ✅                       | ❌    |
| Tail Workers                      | ❌                       | ❌    |
| Real-time logs                    | ✅                       | ✅    |
| **Bindings**                      |                          |       |
| AI                                | ✅                       | ✅    |
| Analytics Engine                  | ✅                       | ✅    |
| Assets                            | ✅                       | ❌    |
| Browser Rendering                 | ✅                       | ✅    |
| D1                                | ✅                       | ✅    |

> | Dispatcher(Workers for Platforms) | ✅                       | ❌    |

Reviewer: ```suggestion
| Dispatcher (Workers for Platforms) | ✅                       | ❌    |
```
Maybe weird one to include here? Since it's more that you use WFP, and then assets is feature, not that dispatcher is feature, and part of either Workers or Pages? (because answer is neither — it's part of WFP)

| Durable Objects                   | ✅                       | ❌    |
| Environment Variables             | ✅                       | ✅    |
| Hyperdrive                        | ✅                       | ❌    |
| KV                                | ✅                       | ✅    |
| mTLS                              | ✅                       | ✅    |
| Queue Producers                   | ✅                       | ✅    |
| Queue Consumers                   | ✅                       | ❌    |
| R2                                | ✅                       | ✅    |
| Rate Limiting                     | ✅                       | ✅    |
| Secrets                           | ✅                       | ✅    |
| Service bindings                  | ✅                       | ✅    |
| Vectorize                         | ✅                       | ✅    |

> | Version Metadata                  | ✅                       | ❌    |

Reviewer: @dcpena or @ToriLindsay can we get bit of help adding links to the products/features above? So that reader can quickly learn what each thing is?

Author: im taking this

> <sup>1</sup> Recommend using a Worker infront of the Worker with assets to run

Reviewer: ```suggestion
<sup>1</sup> We recommend using a Worker in front of the Worker with assets to run
```

Think this needs more explanation in order to make sense to people. Link to Service Bindings?

middleware, connected through a [service
binding](/workers/runtime-apis/bindings/service-bindings/).

> <sup>2</sup> Redirects can be handled in a Worker or using [bulk

Reviewer: ```suggestion
<sup>2</sup> You can handle redirects by adding code to your Worker, or use [bulk
```

redirects](/rules/url-forwarding/bulk-redirects/#_top)(only available if your
Worker is running on a [route or custom
domain](/workers/configuration/routing/)).

> <sup>3</sup> Cannot be enabled right now. Actively working on removing this

Reviewer: suggest reusing other language from limits section, see other comments
